r/src/array__to_vector.cpp

```diff
@@ -345,6 +345,64 @@ class Converter_Dictionary : public Converter {
   }
 };
 
+class Converter_Struct : public Converter {
+ public:
+  explicit Converter_Struct(const ArrayVector& arrays) : Converter(arrays), converters() {
+    auto first_array =
+        internal::checked_cast<arrow::StructArray*>(Converter::arrays_[0].get());
+    int nf = first_array->num_fields();
+    for (int i = 0; i < nf; i++) {
+      converters.push_back(Converter::Make({first_array->field(i)}));
+    }
+  }
+
+  SEXP Allocate(R_xlen_t n) const {
+    // allocate a data frame column to host each array
+    auto first_array =
+        internal::checked_cast<arrow::StructArray*>(Converter::arrays_[0].get());
+    auto type = first_array->struct_type();
+    int nf = first_array->num_fields();
+    Rcpp::List out(nf);
+    Rcpp::CharacterVector colnames(nf);
+    for (int i = 0; i < nf; i++) {
+      out[i] = converters[i]->Allocate(n);
+      colnames[i] = type->child(i)->name();
+    }
+    IntegerVector rn(2);
+    rn[0] = NA_INTEGER;
+    rn[1] = -n;
+    Rf_setAttrib(out, symbols::row_names, rn);
+    Rf_setAttrib(out, R_NamesSymbol, colnames);
+    Rf_setAttrib(out, R_ClassSymbol, Rf_mkString("data.frame"));
+    return out;
+  }
+
+  Status Ingest_all_nulls(SEXP data, R_xlen_t start, R_xlen_t n) const {
+    int nf = converters.size();
+    for (int i = 0; i < nf; i++) {
+      STOP_IF_NOT_OK(converters[i]->Ingest_all_nulls(VECTOR_ELT(data, i), start, n));
+    }
+    return Status::OK();
+  }
+
+  Status Ingest_some_nulls(SEXP data, const std::shared_ptr<arrow::Array>& array,
+                           R_xlen_t start, R_xlen_t n) const {
+    auto struct_array = internal::checked_cast<arrow::StructArray*>(array.get());
+    int nf = converters.size();
+    ArrayVector arrays(nf);
+    STOP_IF_NOT_OK(struct_array->Flatten(default_memory_pool(), &arrays));
```

Comment:
You're effectively doing nothing with `arrays`, is this intended?

Reply:
Thanks. using it now.

Reply:
Could you add a comment why it is necessary? I just realized that this is needed due to the merge of Nulls. Probably worth a unit test.

Reply:
🤔 getting this: 

``` r
library(arrow, warn.conflicts = FALSE)
library(tibble)

tf <- tempfile()
writeLines('
    { "nuf": {} }
    { "nuf": null }
    { "nuf": { "ps": 78.0, "hello": "hi" } }
    { "nuf": { "ps": 90.0, "hello": "bonjour" } }
    { "nuf": { "hello": "ciao" } }
    { "nuf": { "ps": 19 } }
  ', tf)
tab <- read_json_arrow(tf, as_tibble = FALSE)
as_tibble(tab)
#> # A tibble: 6 x 1
#>   nuf$ps $hello 
#>    <dbl> <chr>  
#> 1     NA <NA>   
#> 2     NA <NA>   
#> 3     78 hi     
#> 4     90 bonjour
#> 5     NA ciao   
#> 6     19 <NA>

struct_array <- tab$column(0)$data()$chunk(0)
struct_array$IsNull(0)
#> [1] FALSE
struct_array$IsNull(1)
#> [1] TRUE
struct_array$IsNull(2)
#> [1] FALSE
struct_array$IsNull(3)
#> [1] FALSE
struct_array$IsNull(4)
#> [1] FALSE
struct_array$IsNull(5)
#> [1] FALSE
```

so only `{ "nuf": null }` makes the item a null, each item or both can be independently null without the struct being null. So maybe there's no need to `Flatten()` and I could just use `->field(i)` in the loop as it used to be

Reply:
Don't forget that the array can be produced by an external program, not just JSON.

Reply:
sure, but right now that's my only reproducible way to get one